tokio/src/task/scope/scope.rs

```diff
@@ -0,0 +1,354 @@
+                        }
+                    }
+                } else {
+                    Ok(task.await)
+                }
+            })
+        };
+
+        // Since `Scope` is `Sync` and `Send` cancellations can happen at any time
+        // in case of invalid use. Therefore we only check cancellations once:
+        // After the task has been spawned. Since the cancellation is already set,
+        // we need to wait for the task to complete. Then we panic due to invalid
+        // API usage.
+        if self.scope.is_cancelled() {
+            futures::executor::block_on(async {
+                let _ = child_task.await;
+            });
+            panic!("Spawn on cancelled Scope");
+        }
+
+        ScopedJoinHandle {
+            handle: child_task,
+            phantom: core::marker::PhantomData,
+        }
+    }
+}
+
+/// Defines how a scope will behave if the `Future` it returns get dropped
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ScopeDropBehavior {
+    /// When a scope is dropped while tasks are outstanding, the current thread
+    /// will panic. Since this will not wait for child tasks to complete, the
+    /// child tasks can outlive the parent in this case.
+    Panic,
+    /// When a scope is dropped while tasks are outstanding, the process will be
+    /// aborted.
+    Abort,
+    /// When a scope is dropped while tasks are outstanding, the current thread
+    /// will be blocked until the tasks in the `scope` completed. This option
+    /// is only available in multithreaded tokio runtimes, and is the default there.
+    BlockToCompletion,
+    /// Ignore that the scope got dropped and continue to run the child tasks.
+    /// Choosing this option will break structured concurrency. It is therefore
+    /// not recommended to pick the option.
+    ContinueTasks,
+}
+
+/// Defines how a scope should cancel its child task once the scope is exited
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ScopeCancelBehavior {
+    /// Once the scope is exited, all still running child tasks will get cancelled.
+    /// The cancellation is asynchronous: Tasks will only notice the
+    /// cancellation the next time they are scheduled.
+    /// This option is the default option.
+    CancelChildTasks,
+    /// Child tasks are allowed to continue to run.
+    /// This option should only be chosen if it is either guaranteed that child
+    /// tasks will join on their own, or if the application uses an additional
+    /// mechanism (like a `CancellationToken`) to signal child tasks to return.
+    ContinueChildTasks,
+}
+
+/// Advanced configuration options for `scope`
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct ScopeOptions {
+    /// Whether tasks should be cancelled once the scope is exited
+    pub cancel_behavior: ScopeCancelBehavior,
+    /// How the scope should behave if it gets dropped instead of being `await`ed
+    pub drop_behavior: ScopeDropBehavior,
```

Comment:
To avoid breaking changes, I think we should either add an empty private field here, so the struct has to be constructed like
```rust
ScopeOptions { 
    cancel_behavior: // whatever
    drop_behavior: // whatever
    ..ScopeOptions::default(),
}
```
or replace the public fields with a builder. That way, adding new options isn't a breaking change.

Reply:
I think a builder is the way to go! One open question I had regarding this was whether the builder at the end should directly build the scope or just returns the `Option`.

The first option would then look along:

```rust
ScopeBuilder::new()
    .set_cancel_behavior(ScopeCancelBehavior::Panic)
    .build(|scope| async move {
        scope.spawn(...);
}).await;
```

I think that looks a bit heavy - maybe rather just build the `Options`

Reply:
Hmm, do you think there's a use-case for building multiple scopes with the same `Options`?